### PR TITLE
Refactor the everest run model

### PR DESCRIPTION
**Pull request by verveerpj**
**Issue**
Refactoring of the everest run model code.

**Approach**
_Short description of the approach_

(Screenshot of new behavior in GUI if applicable)

- [ ] PR title captures the intent of the changes, and is fitting for release notes.
- [ ] Added appropriate release note label
- [ ] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [ ] Make sure unit tests pass locally after every commit (`git rebase -i main
      --exec 'pytest tests/ert/unit_tests -n logical -m "not integration_test"'`)

## When applicable
- [ ] **When there are user facing changes**: Updated documentation
- [ ] **New behavior or changes to existing untested code**: Ensured that unit tests are added (See [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)).
- [ ] **Large PR**: Prepare changes in small commits for more convenient review
- [ ] **Bug fix**: Add regression test for the bug
- [ ] **Bug fix**: Create Backport PR to latest release

### CODE REVIEW

File: src/ert/run_models/everest_run_model.py [hunk 1]
```diff
@@ -152,80 +101,50 @@ def __init__(
...
+        self._batch_id: int = 0
+        self._status: SimulationStatus | None = None
 
         super().__init__(
             config,
-            storage,
+            open_storage(config.ens_path, mode="w"),
             config.queue_config,
-            status_queue,
+            queue.SimpleQueue(),
```
Comment (oyvindeide) [inline src/ert/run_models/everest_run_model.py:144]: Think in the next step we want to use this to forward the events from the base run model, so think it might be better to leave it as it was.

File: tests/everest/test_yaml_parser.py [hunk 2]
```diff
@@ -17,9 +17,13 @@ def test_random_seed(random_seed):
     if random_seed:
         config["environment"] = {"random_seed": random_seed}
     ever_config = EverestConfig.with_defaults(**config)
-    assert ever_config.environment.random_seed == random_seed
     ert_config = everest_to_ert_config(ever_config)
-    assert ert_config.random_seed == random_seed
+    if random_seed is None:
+        assert ever_config.environment.random_seed > 0
```
Comment (oyvindeide) [inline tests/everest/test_yaml_parser.py:22]: This seems like a change in behavior? Suggest splitting the changes related to this into a separate commit.

File: tests/everest/test_everest_output.py [hunk 3]
```diff
@@ -52,13 +52,7 @@ async def test_everest_output(copy_mocked_test_data_to_tmp):
...
     initial_files = set(files)
 
-    # Tests in this class used to fail when a callback was passed in
-    # Use a callback just to see that everything works fine, even though
-    # the callback does nothing
-    def useless_cb(*args, **kwargs):
-        pass
-
-    EverestRunModel.create(config, optimization_callback=useless_cb)
+    EverestRunModel.create(config)
```
Comment (oyvindeide) [inline tests/everest/test_everest_output.py:55]: Good change, but think it should in a separate commit with the logic change that enables this.

File: src/ert/run_models/everest_run_model.py [hunk 4]
```diff
@@ -234,29 +153,39 @@ def create(
...
+        config_json = json.dumps(self._everest_config, sort_keys=True, indent=2)
+        return f"EverestRunModel(config={config_json})"
+
     def run_experiment(
         self, evaluator_server_config: EvaluatorServerConfig, restart: bool = False
     ) -> None:
         self.log_at_startup()
         self.restart = restart
-        self.eval_server_cfg = evaluator_server_config
+        self._eval_server_cfg = evaluator_server_config
```
Comment (oyvindeide) [inline src/ert/run_models/everest_run_model.py:188]: Fine to make these private, but would be good to split it into its own commits.

**Review by oyvindeide — COMMENTED** [review]
I think the changes looks good, but I think we should split it into different commits, I had a hard time following some of the changes. For example using pattern matching could be a separate commit, I also added some other examples.

**Comment by verveerpj** [comment]
I will try to split it up.

File: src/ert/run_models/everest_run_model.py [hunk 5]
```diff
@@ -155,11 +154,18 @@ def __init__(
         simulation_callback: SimulationCallback | None,
         optimization_callback: OptimizerCallback | None,
     ):
-        everest_config = self._add_defaults(everest_config)
-
         Path(everest_config.log_dir).mkdir(parents=True, exist_ok=True)
         Path(everest_config.optimization_output_dir).mkdir(parents=True, exist_ok=True)
 
+        assert everest_config.environment is not None
+        logging.getLogger(EVEREST).info(
```
Comment (yngve-sk) [inline src/ert/run_models/everest_run_model.py:111]: Should this logging maybe also be moved to the environment config as it relates to the seed? Semi-related but I think the two logging statements could be merged into one, if it seems sensible to do in this PR

File: src/ert/run_models/everest_run_model.py [hunk 6]
```diff
@@ -146,6 +142,14 @@ def from_seba_optimal_result(
         )
 
 
+class EverestExitCode(IntEnum):
```
Comment (yngve-sk) [inline src/ert/run_models/everest_run_model.py:91]: Makes distinction between runmodel/fm runner and everserver more clear, nice

File: src/ert/run_models/everest_run_model.py [hunk 7]
```diff
@@ -703,3 +716,42 @@ def _handle_errors(
             self._fm_errors[error_hash]["ids"].append(fm_id)
             error_id = self._fm_errors[error_hash]["error_id"]
             fm_logger.error(err_msg.format(error_id, ""))
+
+
+class SimulatorCache:
```
Comment (yngve-sk) [inline src/ert/run_models/everest_run_model.py:748]: Just a general question maybe not related to this PR: Do we still need a simulator cache? It stores some controls and resulting constraints/objectives. I guess also when we have everest storage we could get it from there instead?

File: src/ert/run_models/everest_run_model.py [hunk 8]
```diff
@@ -703,3 +716,42 @@ def _handle_errors(
...
+            list[
+                tuple[
+                    NDArray[np.float64], NDArray[np.float64], NDArray[np.float64] | None
+                ]
+            ],
+        ] = defaultdict(list)
+
+    def add(
+        self,
+        realization_id: int,
```
Comment (yngve-sk) [inline src/ert/run_models/everest_run_model.py:763]: what kind of realization is `realization_id`? Would comment or make more explicit name if possible

File: src/ert/run_models/everest_run_model.py (hunk 5 above)
Comment (verveerpj) [reply src/ert/run_models/everest_run_model.py:111]: I think the logging is appropriate here, since it is basically announcing that we are going to run optimization with e given seed. We could move it to the config, then it would be more like: logging that we modified the configuration to fill in a defaulted value. We generally don't seem to do that.

I will merge them.

File: src/ert/run_models/everest_run_model.py (hunk 7 above)
Comment (verveerpj) [reply src/ert/run_models/everest_run_model.py:746]: Yes, we do need it, some optimizers may repeatedly ask for the same value. It would be a good idea to get is somehow from storage. Let's look into that when the storage has stabilized.

File: src/ert/run_models/everest_run_model.py (hunk 8 above)
Comment (verveerpj) [reply src/ert/run_models/everest_run_model.py:761]: Agreed: `realization_id` -> `realization_name`

Comment (yngve-sk) [reply src/ert/run_models/everest_run_model.py]: In the ERT codebase otherwise it is usually called `real` / `realization` , not saying it is ideal, but maybe we shouldn't add yet another synonym? 😅 Unless this is specific to Everest somehow?

File: src/ert/run_models/everest_run_model.py (hunk 5 above)
Comment (DanSava) [reply src/ert/run_models/everest_run_model.py:111]: :heavy_plus_sign:

File: src/ert/run_models/everest_run_model.py (hunk 8 above)
Comment (verveerpj) [reply src/ert/run_models/everest_run_model.py]: Reason I used name is that they are strings, and I also use sometimes indices to realizations. I will rename things differently, making sure indices are names appropriately, and go to realizations for the names.

File: src/ert/run_models/everest_run_model.py (hunk 7 above)
Comment (verveerpj) [reply src/ert/run_models/everest_run_model.py:746]: The controls are vectors of float values, so not hashable, hence `lru_cache `is not straigtforward to use. They do not have some kind of hashable  ID, I think numpy arrays don't have that.

For each control vector we will have multiple evaluations, i.e. the different realizations, hence the additional use of the realization names in the cache.

File: src/ert/run_models/everest_run_model.py [hunk 9]
```diff
@@ -775,11 +781,9 @@ def add(
         )
 
     def get(
-        self, realization_id: int, controls: NDArray[np.float64]
+        self, realization: int, controls: NDArray[np.float64]
```
Comment (oyvindeide) [inline src/ert/run_models/everest_run_model.py:784]: Because there was some confusion regarding this, would perhaps to add a docstring to this to describe what realization means here?

**Comment by oyvindeide** [comment]
I had one final small comment, but other than that, looks good from my side 👍 Thanks for splitting up, made it much easier to read through for me at least. Just keep the history as is and dont squash on merge.

File: src/ert/run_models/everest_run_model.py (hunk 9 above)
Comment (yngve-sk) [reply src/ert/run_models/everest_run_model.py:784]: ^agree

Comment (verveerpj) [reply src/ert/run_models/everest_run_model.py:791]: Done

**Review by yngve-sk — APPROVED** [review]
LGTM 💯
